File: 3_PCA_AnáliseFatorial_SVD.py
#--------------------------------------------------------------------------------
# Disciplina: Metodos Quantitativos em CC - 2017
# Professor : Luis Enrique Zarate
# Alunos    : Matheus Alcantara
#             Pedro Ruas
#             Marcos W. Rodrigues
# Abordagens: Analise de Componentes Principais (PCA)
#             Analise Fatorial (FA)
#             Singular Value Decomposition (SVD)
#--------------------------------------------------------------------------------

#Documentation PCA: http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html
#Documentation SVD: http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html
#Documentation FA: http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FactorAnalysis.html#sklearn.decomposition.FactorAnalysis.get_covariance

import numpy as np
from sklearn.decomposition import PCA
from sklearn.decomposition import TruncatedSVD, FactorAnalysis
from sklearn.random_projection import sparse_random_matrix

#PCA
#X = sparse_random_matrix(100, 100, density=0.01, random_state=42)
#X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
X = np.random.randint(0, 10, size=(7, 7))
print(" Matriz:")
print (X)



#PCA(copy=True, iterated_power='auto', n_components=2, random_state=None,svd_solver='auto', tol=0.0, whiten=False)


print("\n\n ================================PCA================================")

pca = PCA(n_components=5)
pca.fit(X)

# Principal axes in feature space, representing the directions of maximum variance in the data. The components are sorted by explained_variance_.
print("Componentes PCA")
print (pca.components_)

#The amount of variance explained by each of the selected components.
print ("Variância explicada por componente:")
print(pca.explained_variance_) 

#Percentage of variance explained by each of the selected components.
print ("Porcentagem de variância explicada por componente:")
print(pca.explained_variance_ratio_) 

print ("Variância explicada acumulada:")
print(pca.explained_variance_ratio_.sum()) 

#Perform dimensionality reduction on X.
print("Redução de dimensionalidade na matriz")
print(pca.transform(X))





#Analise fatorial
print("\n\n ================================ANALISE FATORIAL================================")

#X = sparse_random_matrix(10, 10, density=1, random_state=20)
#X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
#print(X)

fa = FactorAnalysis(n_components=5)
fa.fit(X)

print("Componentes FA")
print (fa.components_)

#Compute data covariance with the FactorAnalysis model.
print("Compute data covariance with the FactorAnalysis model")
print (fa.get_covariance())

#The log likelihood at each iteration.
print ("The log likelihood at each iteration:")
print(fa.loglike_) 

#The estimated noise variance for each feature.
print ("The estimated noise variance for each feature:")
print(fa.noise_variance_) 

#Compute the average log-likelihood of the samples
print ("Scores da amostra:")
print(fa.score(X))

#Perform dimensionality reduction on X.
print("Redução de dimensionalidade na matriz")
print(fa.transform(X))



#SVD
print("\n\n ================================SVD================================")

#X = sparse_random_matrix(10, 10, density=1, random_state=20)
#X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
#print(X)

svd = TruncatedSVD(n_components=5, n_iter=7, random_state=20)
svd.fit(X)

print("Componentes SVD")
print (svd.components_)

#The variance of the training samples transformed by a projection to each component.
print ("The variance of the training samples:")
print(svd.explained_variance_) 

#Percentage of variance explained by each of the selected components.
print ("Porcentagem de variância explicada por componente:")
print(svd.explained_variance_ratio_) 

print ("Variância explicada acumulada:")
print(svd.explained_variance_ratio_.sum()) 


#Perform dimensionality reduction on X.
print("Redução de dimensionalidade na matriz")
print(svd.transform(X))

File: 3_PCA_AnáliseFatorial_SVD.R
#--------------------------------------------------------------------------------
# Disciplina: Metodos Quantitativos em CC - 2017
# Professor : Luis Enrique Zarate
# Alunos    : Matheus Alcantara
#             Pedro Ruas
#             Marcos W. Rodrigues
# Abordagens: Análise de Componentes Principais (Principal Component Analysis - PCA)
#             Análise Fatorial (Factor Analysis - FA)
#             Decomposição em valor Singular (Singular Value Decomposition - SVD)
#--------------------------------------------------------------------------------

#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------
# PCA - Análise de Componentes Principais (Principal Component Analysis)
#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------

library(factoextra)
library(devtools) #install.packages("devtools")
#devtools::install_github("kassambara/factoextra")

data(decathlon2)
#Extract only active individuals and variables for principal component analysis:
decathlon2.active <- decathlon2[1:23, 1:10]
head(decathlon2.active[, 1:6])
res.pca <- prcomp(decathlon2.active, scale=TRUE)
#The values returned, by the function prcomp(), are:
names(res.pca)

print(res.pca)

#the standard deviations of the principal components (the square roots of the eigenvalues)
head(res.pca$sdev)
  
#rotation : the matrix of variable loadings (columns are eigenvectors)
head(unclass(res.pca$rotation)[, 1:4])

#center, scale : the centering and scaling used, or FALSE
# Eigenvalues
eig <- (res.pca$sdev)^2
# Variances in percentage
variance <- eig*100/sum(eig)
# Cumulative variances
cumvar <- cumsum(variance)
eig.decathlon2.active <- data.frame(eig=eig, variance=variance, cumvariance=cumvar)
head(eig.decathlon2.active)

summary(res.pca)

fviz_screeplot(res.pca, ncp=10)

head(res.pca)

fviz_pca_var(res.pca)

fviz_pca_var(res.pca, col.var="cos2") +
  scale_color_gradient2(low="white", mid="blue", 
                        high="red", midpoint=0.5) + theme_minimal()
# Contributions of variables on PC1
fviz_pca_contrib(res.pca, choice = "var", axes = 1)

# Contributions of variables on PC2
fviz_pca_contrib(res.pca, choice = "var", axes = 2)

# Total contribution on PC1 and PC2
fviz_pca_contrib(res.pca, choice = "var", axes = 1:2)

fviz_pca_contrib(res.pca, choice = "var", axes = 1, top = 7)

# Control variable colors using their contributions
fviz_pca_var(res.pca, col.var="contrib")

# Change the gradient color
fviz_pca_var(res.pca, col.var="contrib") +
  scale_color_gradient2(low="white", mid="blue", 
                        high="red", midpoint=50) + theme_minimal()

#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------
# Análise Fatorial (Factor Analysis)
#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------

install.packages("psych")
library(psych) #require(psych)
require(foreign)
library( GPArotation )

#Data
#28 variables, and 1000 observations, who were asked to rate themselves on a Lickert scale of 1 to 6. 
#We should be aware of two issues: 
#Lickert scales are known to produce skewed results because of testing fatigue and general boredom, and 
#people are not always very accurate at self-reporting. 
#The last three variables in the data frame report age, education and gender.
data(bfi)
head(bfi)
describe(bfi)

df <- bfi[1:25]
scree(df)
fa.parallel(bfi)

out <- fa(df,6,fm='minres',rotate='oblimin')
print(out$loadings,cut=.2)
fa.diagram( out )
out <- fa(df,5,fm='minres','oblimin')
print(out$loadings,cut=.2)
fa.diagram( out )

#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------
# SVD - Decomposição em valor Singular (Singular Value Decomposition)
#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------

cnut <- read.dta("http://statistics.ats.ucla.edu/stat/data/cerealnut.dta")

#centering the variables
mds.data <- as.matrix(sweep(cnut[, -1], 2, colMeans(cnut[, -1])))
dismat <- dist(mds.data)
mds.m1 <- cmdscale(dismat, k = 8, eig = TRUE)
mds.m1$eig

mds.m1 <- cmdscale(dismat, k = 2, eig = TRUE)
x <- mds.m1$points[, 1]
y <- mds.m1$points[, 2]
plot(x, y)
text(x + 20, y, label=cnut$brand)

# eigenvalues
xx <- svd(mds.data %*% t(mds.data))
xx$d

# coordinates
xxd <- xx$v %*% sqrt(diag(xx$d))
x1 <- xxd[, 1]
y1 <- xxd[, 2]

plot(x1, y1)
text(x1 + 20, y1, label=cnut$brand)

#--------------------------------------------------------------------------------
#--------------------------------------------------------------------------------
